### PR TITLE
Properly handle interrupts and fix non-flushing demux

### PR DESCRIPTION
**Pull request by shadaj**
Properly handle interrupts and fix non-flushing demux

---
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/hydro-project/hydroflow/pull/477).
* #481
* #479
* __->__ #477

### CODE REVIEW

**Comment by shadaj** [comment]
"handles ctrl-c v5 FINAL FINALEST WORKS"

File: hydro_cli_examples/examples/dedalus_sender/main.rs [hunk 1]
```diff
@@ -16,13 +19,16 @@ async fn main() {
     let peers: Vec<u32> = serde_json::from_str(&std::env::args().nth(1).unwrap()).unwrap();
     let broadcast_sink = broadcast_port.take_sink();
 
+    let periodic = IntervalStream::new(tokio::time::interval(std::time::Duration::from_secs(1)));
```
Comment (MingweiSamuel) [inline hydro_cli_examples/examples/dedalus_sender/main.rs:22]: :+1:

File: hydroflow_cli_integration/src/lib.rs [hunk 2]
```diff
@@ -268,27 +267,49 @@ pub struct ConnectedDemux<T: Connected> {
     sink: Option<DynSink<(u32, T::Input)>>,
 }
 
-// a copy of `Drain` that uses `io::Error` instead of `Infallible`
-struct IoErrorDrain<T> {
+#[pin_project]
+struct DemuxDrain<T> {
     marker: PhantomData<T>,
+    #[pin]
+    sinks: HashMap<u32, DynSink<T>>,
```
Comment (MingweiSamuel) [inline hydroflow_cli_integration/src/lib.rs:274]: Seems reasonable, some concern about `dyn Sink` perf if this is on the hot path, would need to check via benchmarking

File: hydroflow_cli_integration/Cargo.toml [hunk 3]
```diff
@@ -11,6 +11,7 @@ async-trait = "0.1"
 bytes = "1.1.0"
 futures = "0.3.26"
 async-recursion = "1"
+pin-project = "1"
```
[the remark this answers is not in the document]
Comment (shadaj) [reply hydroflow_cli_integration/Cargo.toml:14]: Seems to have issues with trait bounds, after I got rid of the `DynSink` use in the demux.

[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply hydroflow_cli_integration/Cargo.toml:14]: Ok fair enough not worth wasting time over

**Pull request merged**